PhysicsTools/PatAlgos/plugins/BaseMVAValueMapProducer.h

```diff
@@ -63,19 +73,47 @@ class BaseMVAValueMapProducer : public edm::stream::EDProducer<> {
       values_.resize(variablesOrder_.size());
       size_t i=0;
       for(const auto & v : variablesOrder_){
-	positions_[v]=i;
-	reader_.AddVariable(v,(&values_.front())+i);
-	i++;
+        positions_[v]=i;
+        if(tmva_) reader_->AddVariable(v,(&values_.front())+i);
+	    i++;
       }
 //      reader_.BookMVA(name_,iConfig.getParameter<edm::FileInPath>("weightFile").fullPath() );
-      reco::details::loadTMVAWeights(&reader_, name_, iConfig.getParameter<edm::FileInPath>("weightFile").fullPath());
-      produces<edm::ValueMap<float>>();
+    if(tmva_)
+    {
+        reco::details::loadTMVAWeights(reader_, name_, weightfilename_);
+    }else if(tf_)    {
+        tensorflow::setLogging("3");
+        graph_=tensorflow::loadGraphDef(weightfilename_);
+        inputTensorName_=iConfig.getParameter<std::string>("inputTensorName");
+        outputTensorName_=iConfig.getParameter<std::string>("outputTensorName");
+        output_names_=iConfig.getParameter<std::vector<std::string>>("outputNames");
+        for(const auto & s : iConfig.getParameter<std::vector<std::string>>("outputFormulas")) { output_formulas_.push_back(StringObjectFunction<std::vector<float>>(s));} 
+        size_t nThreads = iConfig.getParameter<unsigned int>("nThreads");
+        std::string singleThreadPool = iConfig.getParameter<std::string>("singleThreadPool");
+        tensorflow::SessionOptions sessionOptions;
+        tensorflow::setThreading(sessionOptions, nThreads, singleThreadPool);
+        session_ = tensorflow::createSession(graph_, sessionOptions);
+
+    } else  {
+        //FIXME: throw config error
+    }
+   if(tmva_) produces<edm::ValueMap<float>>();
+   else {
+        for(const auto & n : output_names_){
+            produces<edm::ValueMap<float>>(n);
+        }
+   }
 
   }
   ~BaseMVAValueMapProducer() override {}
 
   void setValue(const std::string var,float val) {
-	values_[positions_[var]]=val;
+      if(positions_.find(var)!=positions_.end())
+          values_[positions_[var]]=val;
+      else {
+       //   edm::LogWarning("MVAUnexpectedVariable") << "Variable not expected: " << var ;
+         // std::cout << "Variable not expected: " << var << std::endl;
```

Comment:
shouldn't this be an exception?

Reply:
no, there could be unused variables computed in derived classes that are not used by a specifc training. E.g. you want to test an MVA with 20 variables or a simplified one with 10. The producer may still compute them all.